Add Buildpack environment related content (#754)
buildpacks/docs: content/docs/for-buildpack-authors/how-to/write-buildpacks/specify-env.md

weight=99
+++

> Environment variables are a common way to configure various buildpacks at build-time.

Reviewer: For the content directed at app developers, this sentence is pretty relevant. But buildpacks have other uses for them:
* to configure other buildpacks (at build-time)
* to make executables available on the path
* to configure the app (at runtime)

Author: added this.

<!--more-->

### POSIX Path Variables

> The following layer path environment variables MUST be set by the `lifecycle` during the `build` and `launch` phases in order to make buildpack dependencies accessible.

Reviewer: I'm gonna do a more detailed pass over this, but at first glance it seems that the text (especially where it pulls from the spec) is geared more toward a lifecycle implementor vs a buildpack author. I think we want to see language like, "to add a directory to the $PATH variable for buildpacks that follow, create a `<layers>/bin` directory and place any executables that you want to make available within that directory. The lifecycle will automatically add the directory to the $PATH for buildpacks that follow."

Author: @natalieparellano any idea when you'll have a detailed look at this? Thanks

> During the `build` phase, each variable designated for build MUST contain absolute paths of all previous buildpacks’ `<layers>/<layer>/` directories that are designated for build.

Reviewer: Okay let's start by saying that when the lifecycle runs buildpacks, it strips everything out of the environment that might have been defined on the build-time base image. It only allows a select set of pre-configured environment variables through: https://github.com/buildpacks/lifecycle/blob/a43d5993a4f2cc23c44b6480ba2ab09fe81d57ed/env/build.go#L9-L19

Then, it applies "buildpack defined" environment variables, which is anything that might have been configured by a buildpack that ran earlier during the build phase. Note that buildpacks can't set environment variables for other buildpacks during the detect phase.

If we look at this directory tree:
```
layers/
└── some-buildpack-id
    └── some-build-layer
        ├── bin
        │   └── some-binary
        ├── env
        │   └── SOME_FIRST_VAR
        ├── env.build
        │   └── SOME_SECOND_VAR
        └── lib
            ├── some-shared-library
            └── some-static-library
```

and the table here: https://github.com/buildpacks/spec/blob/main/buildpack.md#layer-paths

* `some-binary` is added to `PATH` for buildpacks that follow
* `some-shared-library` and `some-static-library` are added to `LD_LIBRARY_PATH` and `LIBRARY_PATH`
(I didn't illustrate the modification of `CPATH` and `PKG_CONFIG_PATH`, but hopefully this is enough to paint the picture)

`SOME_FIRST_VAR` takes precedence over `SOME_SECOND_VAR` when applying the environment modification rules, which we should discuss separately.

Reviewer: Note that the above is only relevant for build because the layer "some-build-layer" is a build layer, meaning that some-build-layer.toml has

```
[types]
build = true
```

> When the exported OCI image is launched, each variable designated for launch MUST contain absolute paths of all buildpacks’ `<layers>/<layer>/` directories that are designated for launch.

Reviewer: Let's move this to a separate "run" section, I think that would be clearer

In either case,

- The lifecycle MUST order all `<layer>` paths to reflect the reversed order of the buildpack group.

> - The lifecycle MUST order all `<layer>` paths provided by a given buildpack alphabetically ascending.

Reviewer: Let's make a section for "when multiple buildpacks define the same variable" or something like that. That would be a good introduction to the modification rules.

Author: added

| `CPATH`                                    | `/include`   | header files     | [x]   |        |
| `PKG_CONFIG_PATH`                          | `/pkgconfig` | pc files         | [x]   |        |

The platform SHOULD NOT assume any other base-image-provided environment variables are inherited by the buildpack.

### Build Environment Variables

During the `build` phase, buildpacks MAY write environment variable files to `<layers>/<layer>/env/`, `<layers>/<layer>/env.build/`, and `<layers>/<layer>/env.launch/` directories.

For each `<layers>/<layer>/` designated as a build layer, for each file written to `<layers>/<layer>/env/` or `<layers>/<layer>/env.build/` by `/bin/build`, the `lifecycle` MUST modify an environment variable in subsequent executions of `/bin/build` according to the [environment variable modification rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules).

For each file written to `<layers>/<layer>/env/` or `<layers>/<layer>/env.launch/` by `/bin/build`, the `lifecycle` MUST modify an environment variable during the launch phase according to the modification rules above.

> ### Process Environment Variables

Reviewer 2: Do we have a concept of "Process Environment Variables"?  I'm aware of build-time env vars and runtime env vars.

Reviewer: Yep there are process-specific environment variables. I think if we show a directory tree showing all the different ways of configuring the environment that will make it clearer.

TBA

> ### Runtime Environment Variables

Reviewer: For runtime environment variables, and process variables, we can use this tree:

```
layers/
├── some-buildpack-id
│   └── some-launch-layer
│       ├── bin
│       │   └── some-binary
│       ├── env
│       │   └── SOME_FIRST_VAR
│       ├── env.launch
│       │   ├── SOME_SECOND_VAR
│       │   └── some-process
│       │       └── SOME_THIRD_VAR
│       └── lib
│           ├── some-shared-library
│           └── some-static-library
└── some-launch-layer.toml
```

It works very much like the tree above, except that some-launch-layer.toml has

```
[types]
launch = true
```

Author: added

TBA

> ### Environment Variable Modification Rules

Reviewer: I think some specific examples would help here. We could use this tree:

```
layers/
├── some-buildpack-id
│   ├── some-layer
│   │   └── env
│   │       ├── SOME_VAR # contents foo
│   │       └── SOME_VAR.append # contents :
│   └── some-layer.toml
└── some-other-buildpack-id
    ├── some-layer
    │   └── env
    │       ├── SOME_VAR # contents bar
    │       └── SOME_VAR.append # contents :
    └── some-layer.toml
```

And walk through what would be the final value of `SOME_VAR` for each of `append`, `default`, `override`, and `prepend`. Hopefully the comments above provide enough context to construct these examples, but LMK if you get stuck here.

Author: @natalieparellano what would be the final value of SOME_VAR? not sure I understand this

Reviewer: It would be `foo:bar`

Here is the relevant part of the spec:

> Append
The value of the environment variable MUST be a concatenation of the file contents and the contents of other files representing that environment variable. Within that environment variable value,
> * Earlier buildpacks' environment variable file contents MUST precede later buildpacks' environment variable file contents.

> Delimiter
The file contents MUST be used to delimit any concatenation within the same layer involving that environment variable.

Author: added

Author: done

The full list of convenience directories is summarized in the table below:

| Env Variable                               | Layer Path   | Contents         | Build | Launch |
|--------------------------------------------|--------------|------------------|-------|--------|
| `PATH`                                     | `/bin`       | binaries         | [x]   | [x]    |
| `LD_LIBRARY_PATH`                          | `/lib`       | shared libraries | [x]   | [x]    |
| `LIBRARY_PATH`                             | `/lib`       | static libraries | [x]   |        |
| `CPATH`                                    | `/include`   | header files     | [x]   |        |
| `PKG_CONFIG_PATH`                          | `/pkgconfig` | pc files         | [x]   |        |

* User-provided variables are then applied, meaning that it's possible for the end-user to override buildpack-provided variables.
* Finally, `platform-defined` variables are applied, which eventually override any previous values.

### Preparing the environment at runtime

> For `runtime` and `process` environment variables, the [tree above](#example) is still applicable except that it doesn't include a `build` layer. Instead it has a`launch` layer, i.e., `some-launch-layer.toml` that has

Reviewer: ```suggestion
At runtime, the lifecycle (or rather, the piece of the lifecycle known as the `launcher` that gets included in the application image) prepares the environment for the application process.

For setting up environment variables at runtime, the [tree above](#example) is still applicable except that the layer must be a `launch` layer, i.e., `some-launch-layer.toml` has
```

Author: updated

```yaml

[types]
launch = true

```

> >Note that the `launcher` binary sets up the environment at `runtime`. The `launcher` is found inside the app image at `/cnb/lifecycle/launcher` and is the entrypoint for any `CNB-built` image.

Reviewer: ```suggestion
>Note that the `launcher` binary is found inside the application image at `/cnb/lifecycle/launcher` and is the entrypoint for any CNB-built image.
```

Author: updated

* In all cases, the behavior of the `lifecycle` is governed by the file suffix for `<layers>/<layer>/<env>/SOME_VAR<.suffix>`. The suffix is optional, and the assumed behavior when no suffix is provided is `override`.

>Note that whenever the suffix is `append` or `prepend` an additional file, `<layers>/<layer>/<env>/SOME_VAR.delim`, is needed to specify the delimiter used during concatenation. If no delimiter is provided, none will be used.

To better understand the above modification rules, let's take a look at the tree below:

```text

layers/
├── some-buildpack-id
│   ├── some-layer
│   │   └── env
│   │       ├── SOME_VAR # contents foo
│   │       └── SOME_VAR.append # contents :
│   └── some-layer.toml
└── some-other-buildpack-id
    ├── some-layer
    │   └── env
    │       ├── SOME_VAR # contents bar
    │       └── SOME_VAR.append # contents :
    └── some-layer.toml

```

Assuming that `some-buildpack-id` comes before `some-other-buildpack-id` in the buildpack group, the final value of `SOME_VAR` shown above  would be `foo:bar`.

> Note that the examples shown on this page are relatively simple. It is possible for a buildpack to double-specify the same variable (i.e., within two or more different layers), and for a buildpack to specify a variable for a particular phase (build or launch) when the layer has type both `build = true` and `launch = true`. In these cases, the lifecycle determines the final value of the variable according to a process outlined in the specification.

Reviewer: ```suggestion
Note that the examples shown on this page are relatively simple. It is possible for a buildpack to double-specify the same variable (i.e., within two or more different layers), and for a buildpack to specify a variable for a particular phase (build or launch) when the layer has type both `build = true` and `launch = true`. Additionally, for runtime variables, buildpacks can specify a variable for a particular process.

In these cases, the lifecycle determines the final value of the variable according to a process outlined in the specification.
```

Author: updated

weight=99
+++

Environment variables are a common way to configure buildpacks at build-time and the application at runtime.

<!--more-->

### Preparing the environment at build time

> When the `lifecycle` runs each buildpack, it first tears down any environment variables defined on the `build-time` base image of the environment. It only allows a [specific set](https://github.com/buildpacks/lifecycle/blob/a43d5993a4f2cc23c44b6480ba2ab09fe81d57ed/env/build.go#L9-L19) of pre-configured environment variables through.

Author: updated

> For the detect phase, the lifecycle then applies user-provided environment variables. For more information, see [clearing the buildpack environment](https://buildpacks.io/docs/for-buildpack-authors/how-to/write-buildpacks/clear-env/)

Author: updated; remove the double "see" word

```

With this tree:

* The current buildpack will see `SOME_VAR=foo` in its environment
* The current buildpack will find `some-binary` in `PATH`
* The current buildpack will find `some-static-library` in `LIBRARY_PATH`

Thus, any `<layers>/<layer>/<env>` directory is for setting environment variables directly, and `<layers>/<layer>/<bin>`, `<layers>/<layer>/<lib>`, etc. offer a convenient way to modify `POSIX` path variables.

The full list of convenience directories is summarized in the table below:

| Env Variable                               | Layer Path   | Contents         | Build | Launch |
|--------------------------------------------|--------------|------------------|-------|--------|
| `PATH`                                     | `/bin`       | binaries         | [x]   | [x]    |
| `LD_LIBRARY_PATH`                          | `/lib`       | shared libraries | [x]   | [x]    |
| `LIBRARY_PATH`                             | `/lib`       | static libraries | [x]   |        |
| `CPATH`                                    | `/include`   | header files     | [x]   |        |
| `PKG_CONFIG_PATH`                          | `/pkgconfig` | pc files         | [x]   |        |

> * If `env/SOME_VAR` and `env.build/SOME_VAR` have a conflict, a given procedure applies to figure out who "wins" when applying the environment [modification rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules).

Author: removed

> * `User-defined` variables are then applied; meaning it's possible to override `buildpack-defined` variables.

Author: updated

> * Finally, `platform-defined` variables are applied, which eventually override any previous values.

Author: updated

> ### Preparing the environment at runtime

Reviewer: I think we can flesh out this section a little more

Author: could you elaborate on this? thanks

Reviewer 2: I'm happy to land this change now, and flesh out this discussion in subsequent changes.

Author: Thanks @AidanDelaney

Author: @natalieparellano still waiting for you

For `runtime` and `process` environment variables, the [tree above](#example) is still applicable except that it doesn't include a `build` layer. Instead it has a`launch` layer, i.e., `some-launch-layer.toml` that has

```yaml

[types]
launch = true

```

>Note that the `launcher` binary sets up the environment at `runtime`. The `launcher` is found inside the app image at `/cnb/lifecycle/launcher` and is the entrypoint for any `CNB-built` image.

### When multiple buildpacks define the same variable

When multiple buildpacks define the same variable, the ["environment modification rules"](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules) come into play.

> The lifecycle MUST consider the name of the environment variable to be the name of the file up to the first period (`.`) or to the end of the name if no periods are present. In all cases, file contents MUST NOT be evaluated by a shell or otherwise modified before inclusion in environment variable values.

Author: added

| Suffix     | Modification Behavior                     |
|------------|-------------------------------------------|
| none       | [Override](#override)                     |
| `.append`  | [Append](#append)                         |
| `.default` | [Default](#default)                       |
| `.delim`   | [Delimiter](#delimiter)                   |
| `.override`| [Override](#override)                     |
| `.prepend` | [Prepend](#prepend)                       |

#### Append

The value of the environment variable MUST be a concatenation of the file contents and the contents of other files representing that environment variable.

#### Default

The value of the environment variable MUST only be the file contents if the environment variable is empty.

#### Delimiter

The file contents MUST be used to delimit any concatenation within the same layer involving that environment variable.

#### Override

The value of the environment variable MUST be the file contents.

#### Prepend

The value of the environment variable MUST be a concatenation of the file contents and the contents of other files representing that environment variable.

> To better understand the above modification rules, let's take a look at the tree below,

Reviewer: ```suggestion
To better understand the above modification rules, let's take a look at the tree below:
```

Author: updated

```text

layers/
├── some-buildpack-id
│   ├── some-layer
│   │   └── env
│   │       ├── SOME_VAR # contents foo
│   │       └── SOME_VAR.append # contents :
│   └── some-layer.toml
└── some-other-buildpack-id
    ├── some-layer
    │   └── env
    │       ├── SOME_VAR # contents bar
    │       └── SOME_VAR.append # contents :
    └── some-layer.toml

```

> Assuming that `some-buildpack-id` comes before `some-other-buildpack-id` in the buildpack group, the final value of `SOME_VAR` shown above  would be `foo:bar`

Reviewer: ```suggestion
Assuming that `some-buildpack-id` comes before `some-other-buildpack-id` in the buildpack group, the final value of `SOME_VAR` shown above  would be `foo:bar`.
```

Author: updated

> >For more information on modifying environment variables, see [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules) specification.

Reviewer: ```suggestion
### Further reading

Note that the examples shown on this page are relatively simple. It is possible for a buildpack to double-specify the same variable (i.e., within two or more different layers), and for a buildpack to specify a variable for a particular phase (build or launch) when the layer has type both `build = true` and `launch = true`. In these cases, the lifecycle determines the final value of the variable according to a process outlined in the specification.

>For more information on modifying environment variables, see the [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules) section of the specification.
```

Author: updated

> Please note that when `clear-env` is not set to `true`, the `lifecycle` MUST set user-provided environment variables in the environment of `/bin/detect` or `/bin/build` such that:
> 
> * For layer path environment variables, user-provided values are prepended before any existing values and are delimited by the OS path list separator.
> * For all other environment variables, user-provided values override any existing values.
> * The environment variable prefix `CNB_` is reserved. It MUST NOT be used for environment variables that are not defined in this specification or approved extensions.
> 
> >For more information on clearing user-defined environment variables, see [Clear the buildpack environment](https://buildpacks.io/docs/for-buildpack-authors/how-to/write-buildpacks/clear-env/) documentation.

Reviewer: ```suggestion
```

Author: removed

> ### Further reading

Reviewer: I think we can remove this part

Author: removed

weight=99
+++

> Environment variables are a common way to configure various buildpacks at build-time, make executables available on the `$PATH` environment variable, and configure the app at runtime.

Reviewer: ```suggestion
Environment variables are a common way to configure buildpacks at build-time and the application at runtime.
```

Author: updated

<!--more-->

When `clear-env` is not set to `true`, the `lifecycle` MUST set user-provided environment variables in the environment of `/bin/detect` or `/bin/build` such that:

* For layer path environment variables, user-provided values are prepended before any existing values and are delimited by the OS path list separator.
* For all other environment variables, user-provided values override any existing values.

> * The environment variable prefix `CNB_` is reserved. It MUST NOT be used for environment variables that are not defined in this specification or approved extensions.

Reviewer: Let's move this content somewhere else. It should probably not be the first thing we share.

Author: Moved to the bottom; please take a look. Thanks

> ### POSIX Path Variables

Reviewer: ```suggestion
### Preparing the environment at build time
```

Author: updated

> When the `lifecycle` runs buildpacks, it first tears down anything defined on the `build-time` base image of the environment. It only allows a [specific set](https://github.com/buildpacks/lifecycle/blob/a43d5993a4f2cc23c44b6480ba2ab09fe81d57ed/env/build.go#L9-L19) of pre-configured environment variables through.

Reviewer: ```suggestion
When the lifecycle runs each buildpack, it first tears down any environment variables defined on the `build-time` base image. It only allows a [specific set](https://github.com/buildpacks/lifecycle/blob/a43d5993a4f2cc23c44b6480ba2ab09fe81d57ed/env/build.go#L9-L19) of pre-configured environment variables through.

For the detect phase, the lifecycle then applies user-provided environment variables (for more information, see [clearing the buildpack environment](XXX)).
```

Author: updated

> Then, it applies `buildpack defined` environment variables including anything that might have been configured by a buildpack that ran earlier during the `build` phase.

Reviewer: ```suggestion
For the build phase, the process is more complex. Before applying user-provided environment variables, the lifecycle applies buildpack-provided environment variables, which is anything that a previous buildpack (a buildpack that ran earlier in the build phase) might have configured in its layers directory.
```

Author: updated

>Note that buildpacks cannot set environment variables for other buildpacks during the `detect` phase.

#### Example

> Looking at the following directory tree and table

Reviewer: ```suggestion
Let's look at the following directory tree to see how layers created by a previous buildpack (with id `some-buildpack-id`) would affect the environment for the current buildpack.
```

Author: updated

```text

layers/
└── some-buildpack-id
    ├── some-build-layer
    │   ├── bin
    │   │   └── some-binary
    │   ├── env
    │   │   └── SOME_VAR # contents foo

>     │   ├── env.build
>     │   │   └── SOME_VAR # contents bar

Reviewer: ```suggestion
```

    │   └── lib

>     │       ├── some-shared-library

Reviewer: ```suggestion
```

    │       └── some-static-library

>     └── some-build-layer.toml

Reviewer: ```suggestion
    └── some-build-layer.toml # has build = true in the [types] table
```

Author: updated

> 

Reviewer: Let's add some explanatory text here

Reviewer: With this tree:
* The current buildpack will see `SOME_VAR=foo` in its environment
* The current buildpack will find `some-binary` in `PATH`
* The current buildpack will find `some-static-library` in `LIBRARY_PATH`

Thus, any `<layers>/<layer>/<env>` directory is for setting environment variables directly, and `<layers>/<layer>/<bin>`, `<layers>/<layer>/<lib>`, etc. offer a convenient way to modify POSIX path variables. The full list of convenience directories is summarized in the table below:

Author: added

```

| Env Variable                               | Layer Path   | Contents         | Build | Launch |
|--------------------------------------------|--------------|------------------|-------|--------|
| `PATH`                                     | `/bin`       | binaries         | [x]   | [x]    |
| `LD_LIBRARY_PATH`                          | `/lib`       | shared libraries | [x]   | [x]    |
| `LIBRARY_PATH`                             | `/lib`       | static libraries | [x]   |        |
| `CPATH`                                    | `/include`   | header files     | [x]   |        |
| `PKG_CONFIG_PATH`                          | `/pkgconfig` | pc files         | [x]   |        |

> * To add a directory to the $PATH variable for buildpacks that follow, we need to create a `<layers>/<layer>/bin` directory and add any executables we want to make available within that directory. The `lifecycle` will automatically add the new `bin` directory to the `PATH` for buildpacks that follow.
> * Binaries, such as `some-binary`, are added to `PATH` for buildpacks that follow
> * Shared libraries, such as `Some-shared-library` and `some-static-library`, are added to `LD_LIBRARY_PATH` and `LIBRARY_PATH`

Reviewer: ```suggestion
```

> * If `env/SOME_VAR` and `env.build/SOME_VAR` have a conflict, a given procedure applies to figure out who "wins" when applying the environment [modification rules](#environment-variable-modification-rules).

Reviewer: Maybe we don't need to talk about this part (too complicated, authors can consult the spec). Maybe we should just mention (at the very end) that it's possible to specify "build time only" and "launch time only" env vars (relevant when a layer is of type build AND launch).

* `User-defined` variables are then applied; meaning it's possible to override `buildpack-defined` variables.

> * Finally, `platform-defined` variables are applied, which eventually override any previous values.

Reviewer: Let's move this later

> The above is only relevant for `build` because the layer `some-build-layer` is a build layer, i.e., `some-build-layer.toml` has
> 
> ```yaml
> 
> [types]
> build = true
> 
> ```
> 
> >Note that the platform SHOULD NOT assume any other `base-image` provided environment variables are inherited by the buildpack.

Reviewer: ```suggestion
```

> ### Build Environment Variables
> 
> During the `build` phase, buildpacks MAY write environment variable files to `<layers>/<layer>/env/`, `<layers>/<layer>/env.build/`, and `<layers>/<layer>/env.launch/` directories.
> 
> For each `<layers>/<layer>/` designated as a build layer, for each file written to `<layers>/<layer>/env/` or `<layers>/<layer>/env.build/` by `/bin/build`, the `lifecycle` MUST modify an environment variable in subsequent executions of `/bin/build` according to the [environment variable modification rules](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules).

Reviewer: ```suggestion
```

> ### Process and Runtime Environment Variables

Reviewer: ```suggestion
### Preparing the environment at runtime
```

Reviewer: It's probably worth noting here that it's still the "lifecycle" or rather a CNB component that sets up the environment at runtime, but it's the minimal piece of the lifecycle known as the launcher binary, not the lifecycle binary. This can be found inside the app image at `/cnb/lifecycle/launcher` and is the entrypoint for any CNB-built image.

For `runtime` and `process` environment variables, the [tree above](#example) is still applicable except that it doesn't include a `build` layer. Instead it has a`launch` layer, i.e., `some-launch-layer.toml` that has

```yaml

[types]
launch = true

> ```

Reviewer: Let's move this later

> ### Multiple Buildpacks Defining the Same Environment Variable

Reviewer: ```suggestion
### When multiple buildpacks define the same variable
```

Author: updated

> During the `build` phase, each variable designated for `build` MUST contain absolute paths of all previous buildpacks’ `<layers>/<layer>/` directories that are designated for build.
> 
> When the exported `OCI image` is launched, each variable designated for launch MUST contain absolute paths of all buildpacks’ `<layers>/<layer>/` directories that are designated for launch.
> 
> In either case,
> 
> * The lifecycle MUST order all `<layer>` paths to reflect the reversed order of the buildpack group.
> * The lifecycle MUST order all `<layer>` paths provided by a given buildpack alphabetically ascending.
> * The lifecycle MUST separate each path with the OS path list separator (e.g. `:` on Linux, `;` on Windows).

Reviewer: ```suggestion
```

> #### Environment Variable Modification Rules

Reviewer: ```suggestion
When multiple buildpacks define the same variable, the ["environment modification rules"](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment-variable-modification-rules) come into play.
```

Author: updated

> The lifecycle MUST consider the name of the environment variable to be the name of the file up to the first period (`.`) or to the end of the name if no periods are present. In all cases, file contents MUST NOT be evaluated by a shell or otherwise modified before inclusion in environment variable values.

Reviewer: This is as far as I got in my review. I think we can still re-work this section to be more for a buildpack author (what they can expect to find in their environment when a previous buildpack does something) vs describing the lifecycle's behavior.
